psdisk/psdisk.c

```diff
@@ -0,0 +1,220 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Phoenix Systems disk tool
+ *
+ * Copyright 2020 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+
+#include <stdio.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <string.h>
+#include <ctype.h>
+#include <stdlib.h>
+#include <sys/list.h>
+
+/* TODO: change access to ptable.h */
+#include "../../phoenix-rtos-filesystems/ptable/ptable.h"
+
+
+#define LOG_ERROR(str, ...) do { fprintf(stderr, __FILE__  ":%d error: " str "\n", __LINE__, ##__VA_ARGS__); } while (0)
+
+
+/* TODO: create partitions list */
+struct {
+	uint32_t pCnt;
+	memory_properties_t mem;
+
+	FILE *file;
+} psdisk_common;
+
+
+static void psdisk_help(void)
+{
+	printf("\nUsage:");
+	printf("\n\tpsdisk [options]");
+	printf("\n\nOptions");
+	printf("\n%s %-35s %-30s", "   - i ", "<path-to-image-output>", "= Specify image path");
+	printf("\n%s %-35s %-30s", "   - s ", "<mem-size;sector-size>", "= Specify memory's parameters");
+	printf("\n%s %-35s %-30s", "   - p ", "<name;offset;size;type-hex-value>", "= Specify partition's parameters");
+	printf("\n%s %-35s %-30s", "       ", " ", "   types: raw partition = 0x51,");
+	printf("\n%s %-35s %-30s", "       ", " ", "          meterfs partition = 0x75.");
+	printf("\n%s %-35s %-30s", "   - h ", " ", "= Show help");
+	printf("\n\n");
+}
+
+
+static ssize_t psdisk_writeData(unsigned int addr, const void *buff, size_t bufflen)
+{
+	/* TO DO: add option to move partition table at the end of an image ?? */
+	return fwrite(buff, sizeof(uint8_t), bufflen, psdisk_common.file);
+}
+
+
+static int psdisk_createpTable(void)
+{
+	psdisk_common.mem.write = psdisk_writeData;
+
+	/* TODO: call ptable_writePartitions function */
+
+
+	return 0;
+}
+
+
+static int psdisk_parsePartitionOpt(int optind, char *argv[])
```

Comment:
do not redefine `optind` as local variable, this is prone to errors (applies everywhere)